[SPARK-16792][SQL] Dataset containing a Case Class with a List type causes a CompileException (converting sequence to list)

michalsenkyr opened the pull request:
## What changes were proposed in this pull request?

Added a `to` call at the end of the code generated by `ScalaReflection.deserializerFor` if the requested type is not a supertype of `WrappedArray[_]` that uses `CanBuildFrom[_, _, _]` to convert result into an arbitrary subtype of `Seq[_]`.

Care was taken to preserve the original deserialization where it is possible to avoid the overhead of conversion in cases where it is not needed

`ScalaReflection.serializerFor` could already be used to serialize any `Seq[_]` so it was not altered

`SQLImplicits` had to be altered and new implicit encoders added to permit serialization of other sequence types

Also fixes [SPARK-16815] Dataset[List[T]] leads to ArrayStoreException

## How was this patch tested?
```bash
./build/mvn -DskipTests clean package && ./dev/run-tests
```

Also manual execution of the following sets of commands in the Spark shell:
```scala
case class TestCC(key: Int, letters: List[String])

val ds1 = sc.makeRDD(Seq(
(List("D")),
(List("S","H")),
(List("F","H")),
(List("D","L","L"))
)).map(x=>(x.length,x)).toDF("key","letters").as[TestCC]

val test1=ds1.map{_.key}
test1.show
```

```scala
case class X(l: List[String])
spark.createDataset(Seq(List("A"))).map(X).show
```

```scala
spark.sqlContext.createDataset(sc.parallelize(List(1) :: Nil)).collect
```

After adding arbitrary sequence support also tested with the following commands:

```scala
case class QueueClass(q: scala.collection.immutable.Queue[Int])

spark.createDataset(Seq(List(1,2,3))).map(x => QueueClass(scala.collection.immutable.Queue(x: _*))).map(_.q.dequeue).collect
```

michalsenkyr committed:
Added call to toList if deserializing into List

michalsenkyr committed:
Added unit test

AmplabJenkins commented:
Can one of the admins verify this patch?

michalsenkyr commented:
I would like to add that the conversion is specific to `List[_]`. I can add support for arbitrary sequence types through the use of `CanBuildFrom` if it is desirable.

We can also support `Set`s in this way by serializing into arrays. See SPARK-17414

SparkQA commented:
**[Test build #3488 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3488/consoleFull)** for PR 16240 at commit [`8c15b47`](https://github.com/apache/spark/commit/8c15b475fb053aef19906d6a465309d299ca7b4d).

SparkQA commented:
**[Test build #3488 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3488/consoleFull)** for PR 16240 at commit [`8c15b47`](https://github.com/apache/spark/commit/8c15b475fb053aef19906d6a465309d299ca7b4d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

michalsenkyr committed:
Added support for arbitrary sequence types.
WARNING: Breaks Seq.toDS for Seq[Product]

michalsenkyr committed:
Workaround for Seq.toDS with case classes (supports only Seq)

michalsenkyr commented:
Added support for arbitrary sequences.

Now also Queues, ArrayBuffers and such can be used in datasets (all are serialized into ArrayType).

I had to alter and add new implicit encoders into `SQLImplicits`. The new encoders are for `Seq` with `Product` combination (essentially only `List`) to disambiguate between `Seq` and `Product` encoders.

However, I encountered a problem with implicits. When constructing a complex Dataset using `Seq.toDS` that includes a `Product` (like a case class) and a sequence, the encoder doesn't seem to be created. When constructed with `spark.createDataset` or when transforming an existing dataset, there is no problem.

I added a workaround by defining a specific implicit just for `Seq`s. This makes the problem go away for existing usages, however other collections cannot be constructed by `Seq.toDS` unless `newProductSeqEncoder[A, T]` is created with the correct type parameters.

If anybody knows how to fix this, let me know.

michalsenkyr commented:
Possible optimization: Instead of conversions using `to`, we can use `Builder`s. This way we could get rid of the conversion overhead. This would require adding a new codegen method that would operate similarly to `MapObjects` but use a provided `Builder` to build the collection directly.

I will wait for a response to this PR before attempting any more modifications.

aray commented on sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala:
```diff
@@ -100,31 +100,76 @@ abstract class SQLImplicits {
+    newLongSeqEncoder
+
+  /** @since 2.2.0 */
+  implicit def newDoubleListEncoder[T <: Seq[Double] with Product : TypeTag]: Encoder[T] =
```
Should this be `newDoubleSeqWithProductEncoder`?

michalsenkyr committed:
Fixed incorrect method name

marmbrus commented:
/cc @cloud-fan

cloud-fan commented:
The overall strategy LGTM.

> I had to alter and add new implicit encoders into SQLImplicits. The new encoders are for Seq with Product combination (essentially only List) to disambiguate between Seq and Product encoders.

Does scala have a clear definition for this case? i.e. we have implicit for both type `A` and `B`, given type `A with B`, which implicit will be picked?

For the optimization, we can do it in follow-up.

michalsenkyr commented:
None of them. The compilation will fail. That is why I had to provide those additional implicits.

```
scala> class Test[T]
defined class Test

scala> implicit def test1[T <: Seq[String]]: Test[T] = null
test1: [T <: Seq[String]]=> Test[T]

scala> implicit def test2[T <: Product]: Test[T] = null
test2: [T <: Product]=> Test[T]

scala> def test[T : Test](t: T) = null
test: [T](t: T)(implicit evidence$1: Test[T])Null

scala> test(List("abc"))
<console>:31: error: ambiguous implicit values:
 both method test1 of type [T <: Seq[String]]=> Test[T]
 and method test2 of type [T <: Product]=> Test[T]
 match expected type Test[List[String]]
       test(List("abc"))
```

cloud-fan commented:
How about we assign priority to implicit rules like http://stackoverflow.com/questions/1886953/is-there-a-way-to-control-which-implicit-conversion-will-be-the-default-used ?

I think we should prefer `Seq` encoder over `Product` encoder, for `Seq with Product`

michalsenkyr commented:
I actually read that but IDEA complained when I tried to place the `Product` encoder into a separate trait. So I opted for specificity.
However, I tried it again right now and even though IDEA still complains, scalac compiles it, all the tests pass and it all works. So I will change it.
I will also test whether this fixes the `toDS`-specific problems before pushing the changes.

michalsenkyr committed:
Removed Seq with Product encoders in favor of LowPrioritySQLImplicits trait

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -312,12 +312,46 @@ object ScalaReflection extends ScalaReflection {
+          val cls = mirror.runtimeClass(t.typeSymbol.asClass)
+          import scala.collection.generic.CanBuildFrom
+          import scala.reflect.ClassTag
+          import scala.util.{Try, Success}
```
spark code style discourage the usage of `Try` and `Success`, can you refactor your code a little bit? i.e. move `cls.getDeclaredMethod("canBuildFrom", classOf[ClassTag[_]])` out of the `Invoke` code block.

michalsenkyr replied:
Done. I tried looking up the code style you mentioned, but only found the [Databricks' Scala Code Style Guide](https://github.com/databricks/scala-style-guide#exception-handling-try-vs-try). And that is not mentioned in the Spark docs as far as I know.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala:
```diff
@@ -100,31 +97,36 @@ abstract class SQLImplicits {
 
   /** @since 1.6.1 */
-  implicit def newProductSeqEncoder[A <: Product : TypeTag]: Encoder[Seq[A]] = ExpressionEncoder()
+  implicit def newProductSeqEncoder[A <: Product : TypeTag, T <: Seq[A] : TypeTag]: Encoder[T] =
```
This is my only concern now. Can you provide more details about it?

michalsenkyr replied:
This one is the same as all the other ones, just with Product subclasses. If you were concerned about the `TypeTag` on `A`, it was actually not needed as `T`'s tag already contains all the information. I just tested it to be sure and removed it.

michalsenkyr committed:
Removal of extraneous TypeTag

michalsenkyr committed:
Removed use of Try and refactored code for readability

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala:
```diff
@@ -100,31 +97,36 @@ abstract class SQLImplicits {
 
   /** @since 1.6.1 */
-  implicit def newProductSeqEncoder[A <: Product : TypeTag]: Encoder[Seq[A]] = ExpressionEncoder()
+  implicit def newProductSeqEncoder[A <: Product, T <: Seq[A] : TypeTag]: Encoder[T] =
```
can we just use `newProductSeqEncoder[T <: Seq[Product] : TypeTag]: Encoder[T]` here? Then we don't need to workaround implicit

michalsenkyr replied:
You are right. `Seq` is declared covariant on `T` so it works and solves all the problems I was having. Thanks

michalsenkyr committed:
Removed explicit type bound in newProductSeqEncoder and unneeded resulting workarounds

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetPrimitiveSuite.scala:
```diff
@@ -130,6 +130,30 @@ class DatasetPrimitiveSuite extends QueryTest with SharedSQLContext {
     checkDataset(Seq(Array(Tuple1(1))).toDS(), Array(Tuple1(1)))
   }
 
+  test("arbitrary sequences") {
```
let's also test nested sequences, e.g. `List(Queue(1))`, and sequences inside product, e.g. `List(1) -> Queue(1)`

michalsenkyr replied:
I added some sequence-product combination tests.
Nested sequences were never supported (tried on master and 2.0.2).

michalsenkyr committed:
Sequence-product combination tests

cloud-fan commented:
retest this please

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetPrimitiveSuite.scala:
```diff
@@ -17,10 +17,21 @@
 
 case class IntClass(value: Int)
 
+case class SeqCC(s: Seq[Int])
```
what does `CC` short for? How about `SeqClass`?

cloud-fan commented:
LGTM, please create 2 more tickets for the optimization you metioned in https://github.com/apache/spark/pull/16240#issuecomment-266318016 and the nested custom collection.

SparkQA commented:
**[Test build #70859 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70859/testReport)** for PR 16240 at commit [`efd0801`](https://github.com/apache/spark/commit/efd0801e24088b90c1157de0cb0bfe8159aeaac5).

SparkQA commented:
**[Test build #70859 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70859/testReport)** for PR 16240 at commit [`efd0801`](https://github.com/apache/spark/commit/efd0801e24088b90c1157de0cb0bfe8159aeaac5).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SeqCC(s: Seq[Int])`
  * `case class ListCC(l: List[Int])`
  * `case class QueueCC(q: Queue[Int])`
  * `case class ComplexCC(seq: SeqCC, list: ListCC, queue: QueueCC)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70859/
Test FAILed.

cloud-fan commented:
you need to fix mima:
```
[error]  * method newDoubleSeqEncoder()org.apache.spark.sql.Encoder in class org.apache.spark.sql.SQLImplicits does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SQLImplicits.newDoubleSeqEncoder")
[error]  * method newFloatSeqEncoder()org.apache.spark.sql.Encoder in class org.apache.spark.sql.SQLImplicits does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SQLImplicits.newFloatSeqEncoder")
[error]  * method newByteSeqEncoder()org.apache.spark.sql.Encoder in class org.apache.spark.sql.SQLImplicits does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SQLImplicits.newByteSeqEncoder")
[error]  * method newLongSeqEncoder()org.apache.spark.sql.Encoder in class org.apache.spark.sql.SQLImplicits does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SQLImplicits.newLongSeqEncoder")
[error]  * method newStringSeqEncoder()org.apache.spark.sql.Encoder in class org.apache.spark.sql.SQLImplicits does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SQLImplicits.newStringSeqEncoder")
[error]  * method newIntSeqEncoder()org.apache.spark.sql.Encoder in class org.apache.spark.sql.SQLImplicits does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SQLImplicits.newIntSeqEncoder")
[error]  * method newBooleanSeqEncoder()org.apache.spark.sql.Encoder in class org.apache.spark.sql.SQLImplicits does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SQLImplicits.newBooleanSeqEncoder")
[error]  * method newShortSeqEncoder()org.apache.spark.sql.Encoder in class org.apache.spark.sql.SQLImplicits does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SQLImplicits.newShortSeqEncoder")
```

michalsenkyr committed:
Test case classes renamed

michalsenkyr committed:
Fix MiMa tests

michalsenkyr commented:
Not sure how to run MiMa tests locally so I tried my best to figure out what was necessary. Hope this fixes it.
The downside of the fix is that I had to restore the original methods in `SQLImplicits`. I removed the `implicit` keyword and added deprecation annotations as only the new methods should be used from now on. Old code importing the methods explicitly should be fine now.

marmbrus commented:
For future reference: https://github.com/apache/spark/blob/master/dev/mima (script to run mima)

marmbrus commented:
ok to test

SparkQA commented:
**[Test build #70943 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70943/testReport)** for PR 16240 at commit [`68810c4`](https://github.com/apache/spark/commit/68810c4efb445d237604b661511b161a1c42a9bd).

SparkQA commented:
**[Test build #70943 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70943/testReport)** for PR 16240 at commit [`68810c4`](https://github.com/apache/spark/commit/68810c4efb445d237604b661511b161a1c42a9bd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala:
```diff
@@ -99,33 +96,96 @@ abstract class SQLImplicits {
+   * @since 1.6.1
+   * @deprecated use [[newIntSequenceEncoder]]
+   */
+  def newIntSeqEncoder: Encoder[Seq[Int]] = ExpressionEncoder()
```
oh you doesn't need to do this, just update the `project/MimaExcludes.scala` to add something like `ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SQLImplicits.newDoubleSeqEncoder")`. Please see the history of the `MimaExcludes.scala` to see how others update this file.

marmbrus replied:
Wait, I'm not sure I agree... Do we want to break binary compatibility for libraries that might be using this function?  That could have even been resolved implicitly, so it would be confusing when it breaks.

cloud-fan replied:
ah i see, makes sense

cloud-fan commented:
thanks, merging to master!

cloud-fan commented:
@michalsenkyr please create 2 more tickets for the optimization you metioned in https://github.com/apache/spark/pull/16240#issuecomment-266318016 and the nested custom collection problem.
